test/subtype.jl

```diff
@@ -915,3 +915,59 @@ ftwoparams(::TwoParams{<:Real,<:Real}) = 3
 # supertype operator
 @test !(Int >: Integer)
 @test Integer >: Int
+
+# #2552
+DiagDict{T} = Dict{T,T}
+f2552(::Dict) = 1
+f2552(::DiagDict) = 2
+@test f2552(Dict(1 => "a")) == 1
+@test f2552(Dict(1 => 1)) == 2
+
+# #8625
+CV8625{T<:Real} = Vector{Complex{T}}
+@test isa([2.0+3.0im],CV8625)
+
+# #8915
+mutable struct D8915{T} where T <: Union{Float32,Float64}
+    D8915(a) = 1
+    D8915(a::Int) = 2
+end
+@test D8915{Float64}(1) = 2
+
+# #11407
+f11407{K,V}(::Dict{K,V},::Dict{Any,V}) = 1
+f11407{K,V}(::Dict{K,V},::Dict{K,Any}) = 2
+d11407 = Dict{Any,Any}()
+@test_throws MethodError f11407(d11407,d11407)
+
+# #12814
+abstract type A12814{N, T} end
+struct B12814{N, T} <: A12814{N, T}
+    x::NTuple{N, T}
+end
+Base.call{T <: A12814, X <: Array}(::Type{T}, x::X) = 1
```

Comment:
I saw that when I tried this is in 0.5, but couldn't figure out the transformation (and the warning had disappeared in 0.6?). How should this be spelled now?